### PR TITLE
Port ImageSearchModal to a functional component

### PR DESCRIPTION
**Pull request by jelly**
Port ImageSearchModal to a functional component

    To simplify using context providers port this component to a functional
    component. The only tricky part is that it's not easy to do a state
    update with a callback with `useState`. The recommended way is to use
    `useEffect` but we can't call onSearchTriggered as that also changes the
    state and re-renders causing an infinite loop so we just always pass
    selected registry.

    Apart from the port to a functional component, this now properly
    truncates very long descriptions which take eat up half the search
    results space.

    Fixes a potential race condition when already having
    typed a search entry and then quickly switching the selected registry.

    Fallback registries are now properly shown when no search registry was
    configured in containers.conf.

### CODE REVIEW

File: src/ImageSearchModal.jsx [hunk 1]
```diff
@@ -9,244 +9,214 @@ import { Radio } from "@patternfly/react-core/dist/esm/components/Radio";
...
+                queryRegistries = registries.search;
+            } else {
+                queryRegistries = [searchValue];
+            }
+        }
         // if a user searches for `docker.io/cockpit` let podman search in the user specified registry.
-        if (this.state.imageIdentifier.includes('/')) {
-            registries = [""];
+        if (imageIdentifier.includes('/')) {
+            queryRegistries = [""];
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:67]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230704-165425-9756b240-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 2]
```diff
@@ -9,244 +9,214 @@ import { Radio } from "@patternfly/react-core/dist/esm/components/Radio";
...
-                        let dialogErrorDetail = "";
 
                         for (const result of reply) {
                             if (result.status === "fulfilled") {
                                 results = results.concat(JSON.parse(result.value));
                             } else {
-                                dialogError = _("Failed to search for new images");
-                                dialogErrorDetail = result.reason ? cockpit.format(_("Failed to search for images: $0"), result.reason.message) : _("Failed to search for images.");
+                                setDialogError(_("Failed to search for new images"));
+                                setDialogErrorDetail(result.reason ? cockpit.format(_("Failed to search for images: $0"), result.reason.message) : _("Failed to search for images."));
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:92]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230704-165425-9756b240-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 3]
```diff
@@ -9,244 +9,214 @@ import { Radio } from "@patternfly/react-core/dist/esm/components/Radio";
...
                         }
 
-                        this.setState({
-                            imageList: results || [],
-                            searchInProgress: false,
-                            searchFinished: true,
-                            dialogError,
-                            dialogErrorDetail
-                        });
+                        setImageList(results || []);
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:96]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230704-165425-9756b240-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 4]
```diff
@@ -9,244 +9,214 @@ import { Radio } from "@patternfly/react-core/dist/esm/components/Radio";
...
-        );
-    }
-}
+    };
+
+    const onToggleUser = ev => setIsSystem(ev.currentTarget.value === "system");
+    const onDownloadClicked = () => {
+        const selectedImageName = imageList[selected].Name;
+        if (activeConnection)
+            activeConnection.close();
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:119]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230704-165425-9756b240-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 5]
```diff
@@ -9,244 +9,214 @@ import { Radio } from "@patternfly/react-core/dist/esm/components/Radio";
...
+        Dialogs.close();
+        downloadImage(selectedImageName, imageTag, isSystem);
+    };
+
+    const defaultBody = (
+        <>
+            <Form isHorizontal>
+                {dialogError && <ErrorNotification errorMessage={dialogError} errorDetail={dialogErrorDetail} />}
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:127]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230704-165425-9756b240-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 6]
```diff
@@ -9,244 +9,214 @@ import { Radio } from "@patternfly/react-core/dist/esm/components/Radio";
...
+                                   value={imageIdentifier}
+                                   onKeyPress={onKeyPress}
+                                   onChange={(_event, value) => setImageIdentifier(value)} />
+                    </FormGroup>
+                    <FormGroup fieldId="registry-select" label={_("in")}>
+                        <FormSelect id='registry-select'
+                            value={selectedRegistry}
+                            onChange={(_ev, value) => { setSelectedRegistry(value); onSearchTriggered(value, false) }}>
+                            <FormSelectOption value="" key="all" label={_("All registries")} />
+                            {(registries.search || []).map(r => <FormSelectOption value={r} key={r} label={r} />)}
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:147]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230704-165425-9756b240-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 7]
```diff
@@ -9,244 +9,214 @@ import { Radio } from "@patternfly/react-core/dist/esm/components/Radio";
...
+                        );
+                    })}
+                </DataList>}
+            </>}
+        </>
+    );
+
+    const handleClose = () => {
+        if (activeConnection)
+            activeConnection.close();
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:191]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230704-165425-9756b240-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by jelly** [comment]
Let's also fix.
image

File: src/ImageSearchModal.jsx [hunk 8]
```diff
@@ -15,238 +15,203 @@ import cockpit from 'cockpit';
...
+    const [selectedRegistry, setSelectedRegistry] = useState("");
+    const [selected, setSelected] = useState("");
+    const [dialogError, setDialogError] = useState("");
+    const [dialogErrorDetail, setDialogErrorDetail] = useState("");
+    const [typingTimeout, setTypingTimeout] = useState(null);
+
+    let activeConnection = null;
+    const Dialogs = useDialogs();
+    // Registries to use for searching
+    const searchRegistries = registries.search && registries.length !== 0 ? registries.search : fallbackRegistries;
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:40]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230705-092041-36f39867-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 9]
```diff
@@ -15,238 +15,203 @@ import cockpit from 'cockpit';
...
-        let registries = Object.keys(this.props.registries).length !== 0 ? [this.state.registry] : fallbackRegistries;
+        let queryRegistries = searchRegistries;
+        if (searchRegistry !== "") {
+            queryRegistries = [searchRegistry];
+        }
         // if a user searches for `docker.io/cockpit` let podman search in the user specified registry.
-        if (this.state.imageIdentifier.includes('/')) {
-            registries = [""];
+        if (imageIdentifier.includes('/')) {
+            queryRegistries = [""];
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:67]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230705-092041-36f39867-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 10]
```diff
@@ -15,238 +15,203 @@ import cockpit from 'cockpit';
...
-                        let dialogErrorDetail = "";
 
                         for (const result of reply) {
                             if (result.status === "fulfilled") {
                                 results = results.concat(JSON.parse(result.value));
                             } else {
-                                dialogError = _("Failed to search for new images");
-                                dialogErrorDetail = result.reason ? cockpit.format(_("Failed to search for images: $0"), result.reason.message) : _("Failed to search for images.");
+                                setDialogError(_("Failed to search for new images"));
+                                setDialogErrorDetail(result.reason ? cockpit.format(_("Failed to search for images: $0"), result.reason.message) : _("Failed to search for images."));
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:92]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230705-092041-36f39867-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 11]
```diff
@@ -15,238 +15,203 @@ import cockpit from 'cockpit';
...
                         }
 
-                        this.setState({
-                            imageList: results || [],
-                            searchInProgress: false,
-                            searchFinished: true,
-                            dialogError,
-                            dialogErrorDetail
-                        });
+                        setImageList(results || []);
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:96]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230705-092041-36f39867-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 12]
```diff
@@ -15,238 +15,203 @@ import cockpit from 'cockpit';
...
-        );
-    }
-}
+    };
+
+    const onToggleUser = ev => setIsSystem(ev.currentTarget.value === "system");
+    const onDownloadClicked = () => {
+        const selectedImageName = imageList[selected].Name;
+        if (activeConnection)
+            activeConnection.close();
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:120]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230705-092041-36f39867-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 13]
```diff
@@ -15,238 +15,203 @@ import cockpit from 'cockpit';
...
+        Dialogs.close();
+        downloadImage(selectedImageName, imageTag, isSystem);
+    };
+
+    const handleClose = () => {
+        if (activeConnection)
+            activeConnection.close();
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:127]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230705-092041-36f39867-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 14]
```diff
@@ -15,238 +15,203 @@ import cockpit from 'cockpit';
...
+                   <Button variant='primary' isDisabled={selected === ""} onClick={onDownloadClicked}>
+                       {_("Download")}
+                   </Button>
+                   <Button variant='link' className='btn-cancel' onClick={handleClose}>
+                       {_("Cancel")}
+                   </Button>
+               </>}
+        >
+            <Form isHorizontal>
+                {dialogError && <ErrorNotification errorMessage={dialogError} errorDetail={dialogErrorDetail} />}
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:156]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230705-092041-36f39867-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/ImageSearchModal.jsx [hunk 15]
```diff
@@ -15,238 +15,203 @@ import cockpit from 'cockpit';
...
+                                   value={imageIdentifier}
+                                   onKeyDown={onKeyPress}
+                                   onChange={(_event, value) => setImageIdentifier(value)} />
+                    </FormGroup>
+                    <FormGroup fieldId="registry-select" label={_("in")}>
+                        <FormSelect id='registry-select'
+                            value={selectedRegistry}
+                            onChange={(_ev, value) => { setSelectedRegistry(value); clearTimeout(typingTimeout); onSearchTriggered(value, false) }}>
+                            <FormSelectOption value="" key="all" label={_("All registries")} />
+                            {(searchRegistries || []).map(r => <FormSelectOption value={r} key={r} label={r} />)}
```
Comment (cockpituous) [inline src/ImageSearchModal.jsx:176]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1332-20230705-092041-36f39867-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Pull request merged**